### PR TITLE
CRIU throws UnsupportedOperationException when UnsatisfiedLinkError

### PR DESCRIPTION
**Pull request by JasonFengJ9**
CRIU throws `UnsupportedOperationException` when `UnsatisfiedLinkError` occurred.

`isCRIUSupportEnabled()` returns `false` if the `j9criu29` library wasn't load successfully;
Set `errorMsg` to be returned via `CRIUSupport.getErrorMessage()`.

closes https://github.com/eclipse-openj9/openj9/issues/15554

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
Rather than this approach we need a way to return false in `isCRIUSupportEnabled` if there is a problem loading the library, and we need `getErrorMessage` to reflect that message

**Comment by JasonFengJ9** [comment]
> Rather than this approach we need a way to return false in isCRIUSupportEnabled if there is a problem loading the library, and we need getErrorMessage to reflect that message

`CRIUSupport.isCRIUSupportEnabled()` is to check if `-XX:+EnableCRIUSupport` was specified for JVM. We moved `System.loadLibrary("j9criu29")` after `InternalCRIUSupport.isCheckpointAllowed()` to improve startup performance.
The `UnsatisfiedLinkError` only appears when loading the library or performing checkpoint, added a field `errorMsg` to record those errors.
Please have another look.

**Comment by tajila** [comment]
If a user does `CRIUSupport.isCRIUSupportEnabled()` it would be odd to return an exception that says `UnsupportedOperationException("CRIU support is not enabled")`

In terms of performance, if we called loadNativeLibrary in `CRIUSupport::isCRIUSupportEnabled` it shouldnt affect startup too much since the JCL is using `InternalCRIUSupport::isCRIUSupportEnabled` which doesnt do a library load.

**Comment by JasonFengJ9** [comment]
> If a user does CRIUSupport.isCRIUSupportEnabled() it would be odd to return an exception that says UnsupportedOperationException("CRIU support is not enabled")

It returns `false` and set error message like `CRIU is not enabled - ...`, no exception thrown in this method.

> In terms of performance, if we called loadNativeLibrary in CRIUSupport::isCRIUSupportEnabled it shouldnt affect startup too much since the JCL is using InternalCRIUSupport::isCRIUSupportEnabled which doesnt do a library load.

Moved `init()` into `CRIUSupport::isCRIUSupportEnabled` which returns false if there is a problem loading `j9criu29` library.

File: jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java [hunk 1]
```diff
@@ -176,18 +179,21 @@ public static boolean isCheckpointAllowed() {
...
-				s = "There was a problem loaded the criu native library.\n" //$NON-NLS-1$
-						+ "Please check that criu is installed on the machine by running `criu check`.\n" //$NON-NLS-1$
-						+ "Also, please ensure that the JDK is criu enabled by contacting your JDK provider."; //$NON-NLS-1$
+		if (errorMsg == null) {
+			if (InternalCRIUSupport.isCRIUSupportEnabled()) {
+				if (!nativeLoaded) {
+					errorMsg = "There was a problem loaded the criu native library.\n" //$NON-NLS-1$
+							+ "Please check that criu is installed on the machine by running `criu check`.\n" //$NON-NLS-1$
+							+ "Also, please ensure that the JDK is criu enabled by contacting your JDK provider."; //$NON-NLS-1$
+				} else {
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java]: Removed.

File: jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java [hunk 2]
```diff
@@ -585,24 +591,28 @@ private static RestoreException throwSetEnvException(Throwable cause) {
...
 
-		if (InternalCRIUSupport.isCheckpointAllowed()) {
-			init();
-			checkpointJVMImpl(imageDir, leaveRunning, shellJob, extUnixSupport, logLevel, logFile, fileLocks, workDir,
-					tcpEstablished, autoDedup, trackMemory, unprivileged);
-		} else {
-			if (InternalCRIUSupport.isCRIUSupportEnabled()) {
-				throw new UnsupportedOperationException(
-					"Running in non-portable mode (only one checkpoint is allowed), and we have already checkpointed once"); //$NON-NLS-1$
+			if (InternalCRIUSupport.isCheckpointAllowed()) {
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:594]: Moved those hooks after `isCheckpointAllowed()`.

File: jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java [hunk 3]
```diff
@@ -585,24 +590,27 @@ private static RestoreException throwSetEnvException(Throwable cause) {
...
-			checkpointJVMImpl(imageDir, leaveRunning, shellJob, extUnixSupport, logLevel, logFile, fileLocks, workDir,
-					tcpEstablished, autoDedup, trackMemory, unprivileged);
-		} else {
-			if (InternalCRIUSupport.isCRIUSupportEnabled()) {
-				throw new UnsupportedOperationException(
-					"Running in non-portable mode (only one checkpoint is allowed), and we have already checkpointed once"); //$NON-NLS-1$
+				try {
+					checkpointJVMImpl(imageDir, leaveRunning, shellJob, extUnixSupport, logLevel, logFile, fileLocks,
+							workDir, tcpEstablished, autoDedup, trackMemory, unprivileged);
+				} catch (UnsatisfiedLinkError ule) {
```
Comment (tajila) [inline jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:604]: is this exception still possible if we've loaded the library?

Comment (tajila) [reply jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:604]: perhaps we should throw an internalVMError instead. I think the message should say that there is something wrong with the JDK

**Review by tajila — APPROVED** [review]

File: jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java (hunk 3 above)
Comment (tajila) [reply jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:604]: okay, if we load the library but fail to bind the method then the problem is in the JDK, so I think internalVMError (or something like that) would be more appropriate. The message should indicate that there is a problem with libj9criu in the JDK.

Comment (JasonFengJ9) [reply jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:604]: Missed this notification, `InternalError("There is a problem with libj9criu in the JDK")` is added.

**Comment by tajila** [comment]
Jenkins compile xlinuxcriu jdk11

**Pull request merged**